File: neetcode/go/15.3-sum.go
/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */

// @lc code=start
func threeSum(nums []int) [][]int {
	res := [][]int{}
	sort.Ints(nums)

	for i := 0; i < len(nums)-2; i++ {
		if i == 0 || (i > 0 && nums[i] != nums[i-1]) {
			start := i + 1
			end := len(nums) - 1
			target := -1 * nums[i]

			for start < end {
				if nums[start]+nums[end] == target {
					res = append(res, []int{nums[i], nums[start], nums[end]})

					for start < end && nums[start] == nums[start+1] {
						start++
					}
					for start < end && nums[end] == nums[end-1] {
						end--
					}
					start++
					end--
				} else if nums[start]+nums[end] > target {
					end--
				} else {
					start++
				}
			}
		}
	}
	return res
}

// @lc code=end

File: neetcode/python/15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = set()
        positives = []
        negatives = []
        zeroes = []
        
        for num in nums:
            if num > 0:
                positives.append(num)
            elif num < 0:
                negatives.append(num)
            else:
                zeroes.append(num)

        p_lookup, n_lookup = set(positives), set(negatives)

        if zeroes:
            if len(zeroes) >= 3:
                res.add((0, 0, 0))

            for num in p_lookup:
                neg_num = -1 * num
                if neg_num in n_lookup:
                    res.add((neg_num, 0, num))

        for i in range(len(negatives)):
            for j in range(i+1, len(negatives)):
                pos_num = -1 * (negatives[i] + negatives[j])
                if pos_num in p_lookup:
                    res.add(tuple(sorted((negatives[i], negatives[j], pos_num))))
    
        for i in range(len(positives)):
            for j in range(i+1, len(positives)):
                neg_num = -1 * (positives[i] + positives[j])
                if neg_num in n_lookup:
                    res.add(tuple(sorted((neg_num, positives[i], positives[j]))))
        return res
                            
# @lc code=end

